# Data Manipulation using Python
## Marketing for Financial Services

### Batch C Group 1

In [ ]:
import pandas as pd
import numpy as np

In [111]:
#loading main customer detail csv file
df=pd.read_csv("Customer_and_bank details_p1.csv")
df

Customer_id  age          job  marital            education  default  \
0                1   56     services  married          high.school       no   
1                2   45     services  married             basic.9y  unknown   
2                3   59       admin.  married  professional.course       no   
3                4   41  blue-collar  married              unknown  unknown   
4                5   24   technician   single  professional.course       no   
...            ...  ...          ...      ...                  ...      ...   
37079        37080   73      retired  married  professional.course       no   
37080        37081   46  blue-collar  married  professional.course       no   
37081        37082   56      retired  married    university.degree       no   
37082        37083   44   technician  married  professional.course       no   
37083        37084   74      retired  married  professional.course       no   

      housing loan Region_Code State_Code City_Code  
0          no  yes           3         S1        C1  
1          no   no           3         S1        C1  
2          no   no           4         S2        C2  
3          no   no           3         S3        C3  
4         yes   no           3         S3        C3  
...       ...  ...         ...        ...       ...  
37079     yes   no           2        S16       C21  
37080      no   no           2        S16       C21  
37081     yes   no           2        S16       C21  
37082      no   no           4        S17       C49  
37083     yes   no           1         S6      C113  

[37084 rows x 11 columns]

In [112]:
# loading customer response detail csv file
df1=pd.read_csv("Customer_Response_data_p1.csv")

In [113]:
# merging df and df1
df2=pd.merge(df,df1)
df2

Customer_id  age          job  marital            education  default  \
0                1   56     services  married          high.school       no   
1                2   45     services  married             basic.9y  unknown   
2                3   59       admin.  married  professional.course       no   
3                4   41  blue-collar  married              unknown  unknown   
4                5   24   technician   single  professional.course       no   
...            ...  ...          ...      ...                  ...      ...   
37079        37080   73      retired  married  professional.course       no   
37080        37081   46  blue-collar  married  professional.course       no   
37081        37082   56      retired  married    university.degree       no   
37082        37083   44   technician  married  professional.course       no   
37083        37084   74      retired  married  professional.course       no   

      housing loan Region_Code State_Code City_Code    y  
0          no  yes           3         S1        C1   no  
1          no   no           3         S1        C1   no  
2          no   no           4         S2        C2   no  
3          no   no           3         S3        C3   no  
4         yes   no           3         S3        C3   no  
...       ...  ...         ...        ...       ...  ...  
37079     yes   no           2        S16       C21  yes  
37080      no   no           2        S16       C21   no  
37081     yes   no           2        S16       C21   no  
37082      no   no           4        S17       C49  yes  
37083     yes   no           1         S6      C113   no  

[37084 rows x 12 columns]

In [114]:
# loading customer contact detail csv file
df3=pd.read_csv("Customer_campaign_details_p1.csv")
data=pd.merge(df2,df3)

In [115]:
#loading customer postal code csv file
p=pd.read_csv("Customer_Postal_Code_details.csv")
p=p.rename(columns={'customer_id': 'Customer_id'})
data=pd.merge(data,p)

In [116]:
# main dataframe
data

Customer_id  age          job  marital            education  default  \
0                1   56     services  married          high.school       no   
1                2   45     services  married             basic.9y  unknown   
2                3   59       admin.  married  professional.course       no   
3                4   41  blue-collar  married              unknown  unknown   
4                5   24   technician   single  professional.course       no   
...            ...  ...          ...      ...                  ...      ...   
37079        37080   73      retired  married  professional.course       no   
37080        37081   46  blue-collar  married  professional.course       no   
37081        37082   56      retired  married    university.degree       no   
37082        37083   44   technician  married  professional.course       no   
37083        37084   74      retired  married  professional.course       no   

      housing loan Region_Code State_Code  ...    y    contact month  \
0          no  yes           3         S1  ...   no  telephone   may   
1          no   no           3         S1  ...   no  telephone   may   
2          no   no           4         S2  ...   no  telephone   may   
3          no   no           3         S3  ...   no  telephone   may   
4         yes   no           3         S3  ...   no  telephone   may   
...       ...  ...         ...        ...  ...  ...        ...   ...   
37079     yes   no           2        S16  ...  yes   cellular   nov   
37080      no   no           2        S16  ...   no   cellular   nov   
37081     yes   no           2        S16  ...   no   cellular   nov   
37082      no   no           4        S17  ...  yes   cellular   nov   
37083     yes   no           1         S6  ...   no   cellular   nov   

      day_of_week duration  campaign  pdays  previous     poutcome Postal Code  
0             mon      307         1    999         0  nonexistent       42420  
1             mon      198         1    999         0  nonexistent       42420  
2             mon      139         1    999         0  nonexistent       90036  
3             mon      217         1    999         0  nonexistent       33311  
4             mon      380         1    999         0  nonexistent       33311  
...           ...      ...       ...    ...       ...          ...         ...  
37079         fri      334         1    999         0  nonexistent       10009  
37080         fri      383         1    999         0  nonexistent       10011  
37081         fri      189         2    999         0  nonexistent       10009  
37082         fri      442         1    999         0  nonexistent       85254  
37083         fri      239         3    999         1      failure       79109  

[37084 rows x 21 columns]

In [117]:
# Finding the name of the columns
data.columns

Index(['Customer_id', 'age', 'job', 'marital', 'education', 'default',
       'housing', 'loan', 'Region_Code', 'State_Code', 'City_Code', 'y',
       'contact', 'month', 'day_of_week', 'duration', 'campaign', 'pdays',
       'previous', 'poutcome', 'Postal Code'],
      dtype='object')

In [118]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 37084 entries, 0 to 37083
Data columns (total 21 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Customer_id  37084 non-null  int64 
 1   age          37084 non-null  int64 
 2   job          37084 non-null  object
 3   marital      37084 non-null  object
 4   education    37084 non-null  object
 5   default      37084 non-null  object
 6   housing      37084 non-null  object
 7   loan         37084 non-null  object
 8   Region_Code  37084 non-null  object
 9   State_Code   37084 non-null  object
 10  City_Code    37084 non-null  object
 11  y            37084 non-null  object
 12  contact      37084 non-null  object
 13  month        37084 non-null  object
 14  day_of_week  37084 non-null  object
 15  duration     37084 non-null  int64 
 16  campaign     37084 non-null  int64 
 17  pdays        37084 non-null  int64 
 18  previous     37084 non-null  int64 
 19  poutcome     37084 non-nu

###### Initial Findings of The Data:
###### There are total of 37084 records
###### 6 numeric attributes , 14 are categorical attribute
###### 1 target attribute y

### Q1) Percentage turnout of marketing campaign

In [119]:
y = df1.groupby('y').count().rename(columns={'Customer_id':'Number_of_Response'})
y['percentage_of_response'] = round((y['Number_of_Response']/(y['Number_of_Response'].sum()))*100,4)
y

Number_of_Response  percentage_of_response
y                                              
no                32876                 88.6528
yes                4208                 11.3472

##### 11.3472% of the contacted customer converted as Investor 

### Q2) Right mode to contact the customers (telephone or mobile)

In [120]:

## Subscription Rate by Contact

con=pd.crosstab(data['y'],data['contact'])
print(con,"\n")
con=pd.crosstab(data['y'],data['contact']).apply(lambda x: x/x.sum() * 100)
print(con)

contact  cellular  telephone
y                           
no          20030      12846
yes          3492        716 

contact   cellular  telephone
y                            
no       85.154324  94.720543
yes      14.845676   5.279457


##### Contact Type: Customers who are contacted by cellphone are more likely to subscribe to the term deposit than telephone. As only 5.2795% of the customer contacted through telephone has agreed to become Investor against 14.8457% by cell phone. Hence we can conclude that cell(mobile) is the right mode for the success of campaign

### Q3) Analysis on attempts made to turn a person into successful depositor

In [121]:
data['Response_in_Binary'] = np.where(data['y'] == 'no', 0 ,1)
cam = data[['Customer_id','Response_in_Binary','y','campaign']].groupby('campaign').agg({'Response_in_Binary': np.sum}).rename(columns={'Response_in_Binary': 'No_of_Customer_Converted'})
cam1 = data[['Customer_id','campaign']].groupby(['campaign']).count().rename(columns={'Customer_id' : 'Number_of_Customer_for_given_no_of_attempt'})
cam2 = pd.concat([cam1,cam], axis = 1)
cam2.reset_index(inplace=True)
cam2.rename(columns={'campaign' : 'No_of_attempt_made_on_a_customer'},inplace=True)
cam2["Percentage_conversion"] = round(cam2['No_of_Customer_Converted']/cam2['Number_of_Customer_for_given_no_of_attempt']*100,4)


In [122]:
g1 = cam2[cam2['No_of_attempt_made_on_a_customer']<9]
g2 = cam2[cam2['No_of_attempt_made_on_a_customer'].isin(range(9,18))].reset_index(drop=True)
g3 = cam2[cam2['No_of_attempt_made_on_a_customer']>17].reset_index(drop=True)

In [123]:
h1 = pd.DataFrame(g1[['Number_of_Customer_for_given_no_of_attempt','No_of_Customer_Converted']].agg({'Number_of_Customer_for_given_no_of_attempt':np.sum, 'No_of_Customer_Converted': np.sum})).transpose()
h1.insert(loc=0,column='Attempt Range', value='1 to 8')
h2 = pd.DataFrame(g2[['Number_of_Customer_for_given_no_of_attempt','No_of_Customer_Converted']].agg({'Number_of_Customer_for_given_no_of_attempt':np.sum, 'No_of_Customer_Converted': np.sum})).transpose()
h2.insert(loc=0,column='Attempt Range', value='9 to 17')
h3 = pd.DataFrame(g3[['Number_of_Customer_for_given_no_of_attempt','No_of_Customer_Converted']].agg({'Number_of_Customer_for_given_no_of_attempt':np.sum, 'No_of_Customer_Converted': np.sum})).transpose()
h3.insert(loc=0,column='Attempt Range', value='more than 17')
df_5 = pd.concat([h1,h2,h3], axis=0).reset_index(drop = True)
df_5["Percentage conversion to investor"] = round(df_5['No_of_Customer_Converted']/df_5['Number_of_Customer_for_given_no_of_attempt']*100,4)
df_5

Attempt Range  Number_of_Customer_for_given_no_of_attempt  \
0        1 to 8                                       35835   
1       9 to 17                                        1024   
2  more than 17                                         225   

   No_of_Customer_Converted  Percentage conversion to investor  
0                      4160                            11.6088  
1                        47                             4.5898  
2                         1                             0.4444

##### There is no significant change in the no. of Investor if the large number of attempt is made on a single customer. It would be more effective if those large attempts are distributed among more number of customer than on a single customer.

In [124]:
Total_no_of_attempts_made = df3['campaign'].sum()
Total_no_of_customer_converted_as_depositor = cam2['No_of_Customer_Converted'].sum()
d1 = {'Total_no_of_attempts_made': df3['campaign'].sum(),
     'Total_no_of_customer_converted_as_depositor' : data['Response_in_Binary'].sum()}
df6 = pd.DataFrame(d1,index=[0])
df6['Percentage_of_successful_attempt'] = round(df6['Total_no_of_customer_converted_as_depositor']/df6['Total_no_of_attempts_made']*100,4)
df6

Total_no_of_attempts_made  Total_no_of_customer_converted_as_depositor  \
0                      95289                                         4208   

   Percentage_of_successful_attempt  
0                             4.416

##### Only 4.416% of the total attempt is successful.

### Q4) Personal data analysis on marital status, existing loans, education, profession etc. and its impact on the campaign’s success.

In [125]:
#calculate the percentage of each marital status category. 
data.marital.value_counts(normalize= True)

married     0.606164
single      0.280633
divorced    0.111261
unknown     0.001942
Name: marital, dtype: float64

##### In our dataset 60% are married people and 28% people are single.

In [126]:
# Subscription Rate by Marital Status
data = data.loc[(data['marital'] !=  'unknown')]
marital_response_pct = pd.crosstab(data['y'],data['marital']).apply(lambda x: x/x.sum() * 100)
marital_response_pct

marital   divorced    married     single
y                                       
no       89.650994  89.803817  85.807629
yes      10.349006  10.196183  14.192371

####  single's marital status are more tend to take the term deposit and lowest one is married

In [127]:
#calculate the percentage of each job category. 
data.job.value_counts(normalize= True)

admin.           0.254161
blue-collar      0.224252
technician       0.164406
services         0.095699
management       0.071193
retired          0.041500
entrepreneur     0.035313
self-employed    0.034367
housemaid        0.025505
unemployed       0.024884
student          0.020696
unknown          0.008024
Name: job, dtype: float64

#### In our dataset 25% people are admin and 22% people are blue-collar, it means bank targeting high earning people.


In [128]:
# Subscription Rate by job
data = data.loc[(data['job'] !=  'unknown')]
job_response_pct = pd.crosstab(data['y'],data['job']).apply(lambda x: x/x.sum() * 100)
job_response_pct


job     admin.  blue-collar  entrepreneur  housemaid  management    retired  \
y                                                                             
no   86.775805    93.096386      91.58378  90.148305   88.918406  74.348958   
yes  13.224195     6.903614       8.41622   9.851695   11.081594  25.651042   

job  self-employed   services    student  technician  unemployed  
y                                                                 
no       89.701258  91.586674  68.668407   89.235826   85.450597  
yes      10.298742   8.413326  31.331593   10.764174   14.549403

#### job type with student, retired, unemployed are more tend to take term deposit

In [129]:
#calculate the percentage of each education category. 
data.education.value_counts(normalize= True)

university.degree      0.296963
high.school            0.231513
basic.9y               0.146670
professional.course    0.128803
basic.4y               0.101076
basic.6y               0.055754
unknown                0.038785
illiterate             0.000436
Name: education, dtype: float64

#### In our dataset 29.5% people have university degree and 23% people are from high school, more count in university.degree people of course High salary people should have university degree expected. And illiterate count is very less.

In [130]:
## Subscription Rate by Education
data = data.loc[(data['education'] !=  'unknown')&(data['education'] !=  'illiterate')]
edu_response_pct = pd.crosstab(data['y'],data['education']).apply(lambda x: x/x.sum() * 100)
edu_response_pct

education   basic.4y   basic.6y   basic.9y  high.school  professional.course  \
y                                                                              
no         89.571544  91.695164  92.256267    88.917647            88.581095   
yes        10.428456   8.304836   7.743733    11.082353            11.418905   

education  university.degree  
y                             
no                 86.324865  
yes                13.675135

##### education type with university degree, professional course and high school are more tend to take term deposit

In [131]:
## Subscription Rate by housing loan
data = data.loc[(data['housing'] !=  'unknown')]
housing_response_pct = pd.crosstab(data['y'],data['housing']).apply(lambda x: x/x.sum() * 100)
housing_response_pct

housing         no        yes
y                            
no       89.155036  88.481054
yes      10.844964  11.518946

##### From the above DataFrame we can conclude that it is slight easier to convert person having house loan to investor because the interest rate of house loan is less and person prefer to invest in Term deposit to earn more profit.

In [132]:
## Subscription Rate by personal loan
data = data.loc[(data['loan'] !=  'unknown')]
loan_response_pct = pd.crosstab(data['y'],data['loan']).apply(lambda x: x/x.sum() * 100)
loan_response_pct

loan         no        yes
y                         
no    88.702684  89.277736
yes   11.297316  10.722264

##### From the above DataFrame we can conclude that it is slight easier to convert person not having personal loan to investor because the interest rate of personal loan is more and person prefer to pay the loan instaed of investing in Term deposit.

###  Q5) Socio-economical analysis of the customers
#### (i) Consumer price index - monthly indicator 

In [133]:
sc=pd.read_csv('Customer_social_economic_data_p1.csv')
sc
sc.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 37084 entries, 0 to 37083
Data columns (total 6 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Customer_id     37084 non-null  int64  
 1   emp.var.rate    37084 non-null  float64
 2   cons.price.idx  37084 non-null  float64
 3   cons.conf.idx   37084 non-null  float64
 4   euribor3m       37084 non-null  float64
 5   nr.employed     37084 non-null  float64
dtypes: float64(5), int64(1)
memory usage: 1.7 MB


In [134]:
sc1 = pd.merge(sc,data, on='Customer_id')
sc2 = sc1.groupby(['month']).agg({'cons.price.idx':np.mean})
sc2.reset_index(inplace=True)
dr1 = {'jan': 1,'feb': 2,'mar': 3,'apr': 4,'may': 5,'jun': 6,'jul': 7,'aug': 8,'sep':9,'oct': 10,'nov': 11,'dec': 12}
sc2['month in integer'] = sc2['month'].apply(lambda x:dr1[x])
sc2.sort_values('month in integer').reset_index(drop=True)

month  cons.price.idx  month in integer
0   mar       93.085315                 3
1   apr       93.117488                 4
2   may       93.518696                 5
3   jun       94.253363                 6
4   jul       93.895528                 7
5   aug       93.316509                 8
6   sep       93.364324                 9
7   oct       93.178866                10
8   nov       93.201743                11
9   dec       92.715651                12

From the above DataFrame we can conclude that, Consumer price index increases from march to june hence the inflation increases, it goes down from june to august and have only slight fluctuation from august to november and finally in month of december decreases significantly which result in the more purchasing power of customer and hence the success of campaign.
Customers are dominated by group with consumer price index as 93.994


#### (ii) Consumer confidence index - monthly indicator 

In [135]:
sc3 = sc1.groupby(['month']).agg({'cons.conf.idx':np.mean})
sc3.reset_index(inplace=True)
sc3['month in integer'] = sc3['month'].apply(lambda x:dr1[x])
sc3.sort_values('month in integer').reset_index(drop=True)

month  cons.conf.idx  month in integer
0   mar     -42.997753                 3
1   apr     -46.312018                 4
2   may     -40.668031                 5
3   jun     -41.580992                 6
4   jul     -42.374735                 7
5   aug     -35.617782                 8
6   sep     -33.968680                 9
7   oct     -34.613217                10
8   nov     -41.269733                11
9   dec     -33.795205                12

From the above DataFrame we can conclude that over the year Consumer confidence index increses means that customer believe that economy will grow and there job will be secure hence they will invest much which result in the success of campaign.

#### (iii) Euribor 3 month rate - daily indicator 

In [136]:
sc4 = sc1.groupby(['month','day_of_week']).agg({'euribor3m':np.mean})
sc4.reset_index(inplace=True)
dr2 = {'mon':1,'tue':2,'wed':3,'thu':4,'fri':5,'sat':6,'sun':7}
sc4['month in integer'] = sc4['month'].apply(lambda x:dr1[x])
sc4['Week day in integer'] = sc4['day_of_week'].apply(lambda x:dr2[x])
sc4.sort_values(by=['month in integer','Week day in integer']).reset_index(drop = True)

month day_of_week  euribor3m  month in integer  Week day in integer
0    mar         mon   1.252496                 3                    1
1    mar         tue   1.166609                 3                    2
2    mar         wed   1.175952                 3                    3
3    mar         thu   1.169208                 3                    4
4    mar         fri   1.144231                 3                    5
5    apr         mon   1.355988                 4                    1
6    apr         tue   1.333723                 4                    2
7    apr         wed   1.400826                 4                    3
8    apr         thu   1.348277                 4                    4
9    apr         fri   1.384699                 4                    5
10   may         mon   3.102068                 5                    1
11   may         tue   3.423829                 5                    2
12   may         wed   3.338784                 5                    3
13   may         thu   3.094318                 5                    4
14   may         fri   3.332911                 5                    5
15   jun         mon   4.185464                 6                    1
16   jun         tue   3.911717                 6                    2
17   jun         wed   4.202491                 6                    3
18   jun         thu   4.471542                 6                    4
19   jun         fri   4.545600                 6                    5
20   jul         mon   4.789994                 7                    1
21   jul         tue   4.673569                 7                    2
22   jul         wed   4.718125                 7                    3
23   jul         thu   4.656636                 7                    4
24   jul         fri   4.571182                 7                    5
25   aug         mon   4.369040                 8                    1
26   aug         tue   4.338621                 8                    2
27   aug         wed   4.386300                 8                    3
28   aug         thu   4.341464                 8                    4
29   aug         fri   4.221977                 8                    5
30   sep         mon   0.847312                 9                    1
31   sep         tue   0.818558                 9                    2
32   sep         wed   0.839158                 9                    3
33   sep         thu   0.859242                 9                    4
34   sep         fri   0.813263                 9                    5
35   oct         mon   1.331905                10                    1
36   oct         tue   1.348500                10                    2
37   oct         wed   1.222396                10                    3
38   oct         thu   1.014659                10                    4
39   oct         fri   1.313948                10                    5
40   nov         mon   3.592860                11                    1
41   nov         tue   3.719980                11                    2
42   nov         wed   3.782424                11                    3
43   nov         thu   3.838612                11                    4
44   nov         fri   3.723432                11                    5
45   dec         mon   0.917100                12                    1
46   dec         tue   0.709250                12                    2
47   dec         wed   0.846261                12                    3
48   dec         thu   0.859237                12                    4
49   dec         fri   1.107619                12                    5

#### From march till August Euribor rate increses means high interest rate for lending between the bank which is the indicator of economy slow down hence it will negatively affect the campaign. But it suddenly drop in september and then again increases till november finally it decreses in the end of year.

#### (iv) Number of employees - quarterly indicator 

In [137]:
sc5 = sc1.groupby(['month']).agg({'nr.employed':np.mean})
sc5.reset_index(inplace=True)
sc5['month in integer'] = sc5['month'].apply(lambda x:dr1[x])

In [138]:
quarter1 = sc5[sc5['month in integer']<4].agg({'nr.employed':np.mean})
quarter1['Quarter']='Quarter 1'
q1 = pd.DataFrame(quarter1).transpose()
quarter2 = sc5[sc5['month in integer'].isin(range(4,7))].agg({'nr.employed':np.mean})
quarter2['Quarter']='Quarter 2'
q2 = pd.DataFrame(quarter2).transpose()
quarter3 = sc5[sc5['month in integer'].isin(range(7,10))].agg({'nr.employed':np.mean})
quarter3['Quarter']='Quarter 3'
q3 = pd.DataFrame(quarter3).transpose()
quarter4 = sc5[sc5['month in integer']>9].agg({'nr.employed':np.mean})
quarter4['Quarter']='Quarter 4'
q4 = pd.DataFrame(quarter4).transpose()
sc6 = pd.concat([q1,q2,q3,q4], axis=0).reset_index(drop = True)
sc6

nr.employed    Quarter
0  5057.455056  Quarter 1
1  5146.692612  Quarter 2
2  5134.765376  Quarter 3
3   5075.77238  Quarter 4

#### The mean number of employees remains constant (approximately) over the different quarter.



#### (v) Employment variation rate - quarterly indicator

In [139]:
sc['emp.var.rate'].unique()

array([ 1.1,  1.4, -0.1, -0.2, -1.8, -2.9, -3.4, -3. , -1.7, -1.1])

In [140]:
emp = pd.crosstab(index = sc["emp.var.rate"], columns = 'Sum')
emp['Persentase'] = ''
persen = []
for a in emp["Sum"]:
    persen.append(a/emp['Sum'].sum()*100)
emp['Persentase'] = persen
emp

col_0           Sum  Persentase
emp.var.rate                   
-3.4            948    2.556359
-3.0            157    0.423363
-2.9           1484    4.001726
-1.8           8295   22.368137
-1.7            696    1.876820
-1.1            568    1.531658
-0.2              9    0.024269
-0.1           3329    8.976917
 1.1           7000   18.876065
 1.4          14598   39.364686

#### Customers are dominated by group with employee variation rate as 1.4

### Q6) Demographic analysis of the marketing campaign using the master files.

In [141]:
data1=pd.read_csv("Customer_and_bank details_p1.csv")
c=pd.read_csv('City_Master.csv')
d1=pd.merge(data1,c)

In [142]:
d1=pd.merge(d1,df1)

In [143]:
d1.drop(df.columns[[1,2,3, 4,5,6,7]], axis = 1, inplace = True)

In [144]:
d1

Customer_id Region_Code State_Code City_Code  City_Name   y
0                1           3         S1        C1  Henderson  no
1                2           3         S1        C1  Henderson  no
2              539           3         S1        C1  Henderson  no
3              540           3         S1        C1  Henderson  no
4              997           3         S1        C1  Henderson  no
...            ...         ...        ...       ...        ...  ..
32985        19864           2        S25      C531      Mason  no
32986        19865           2        S25      C531      Mason  no
32987        29857           2        S25      C531      Mason  no
32988        29858           2        S25      C531      Mason  no
32989        29859           2        S25      C531      Mason  no

[32990 rows x 6 columns]

In [145]:
r=pd.read_csv("Customer_Postal_Code_details.csv")
r.rename(columns = {'customer_id':'Customer_id'}, inplace = True)

In [146]:
d1=pd.merge(d1,r)
d1

Customer_id Region_Code State_Code City_Code  City_Name   y  \
0                1           3         S1        C1  Henderson  no   
1                2           3         S1        C1  Henderson  no   
2              539           3         S1        C1  Henderson  no   
3              540           3         S1        C1  Henderson  no   
4              997           3         S1        C1  Henderson  no   
...            ...         ...        ...       ...        ...  ..   
32985        19864           2        S25      C531      Mason  no   
32986        19865           2        S25      C531      Mason  no   
32987        29857           2        S25      C531      Mason  no   
32988        29858           2        S25      C531      Mason  no   
32989        29859           2        S25      C531      Mason  no   

       Postal Code  
0            42420  
1            42420  
2            42420  
3            42420  
4            42420  
...            ...  
32985        45040  
32986        45040  
32987        45040  
32988        45040  
32989        45040  

[32990 rows x 7 columns]

In [147]:
analysis_region = pd.pivot_table(d1, index=['Region_Code'], values=['Customer_id'], columns=['y'], aggfunc='count')
analysis_region = analysis_region.reindex(analysis_region['Customer_id'].sort_values(by='yes', ascending=False).index)
analysis_region['total'] = analysis_region['Customer_id']['yes']+analysis_region['Customer_id']['no']
analysis_region.head()

Customer_id        total
y                    no   yes       
Region_Code                         
4                 10158  1329  11487
2                  8208  1006   9214
1                  6692   824   7516
3                  4157   564   4721
Na                   45     7     52

####  People from West Region took up the highest number of  subscription

In [148]:
analysis_state = pd.pivot_table(d1, index=['State_Code'], values=['Customer_id'], columns=['y'], aggfunc='count')
analysis_state = analysis_state.reindex(analysis_state['Customer_id'].sort_values(by='yes', ascending=False).index)
analysis_state['total'] = analysis_state['Customer_id']['yes']+analysis_state['Customer_id']['no']
analysis_state.head()

Customer_id      total
y                   no  yes      
State_Code                       
S2                6429  839  7268
S16               3502  457  3959
S6                3062  373  3435
S5                1632  230  1862
S10               1938  206  2144

#### People from California state  took up the highest number of  subscription 


In [149]:
d1['City_Name'].value_counts().max()

3422

In [150]:
d1['City_Name'].value_counts().min()

3

In [151]:
d2 = pd.DataFrame(d1[['City_Name','Customer_id']].groupby('City_Name').count().reset_index())

In [152]:
print(d2[d2['Customer_id']==3422],"\n")
print(d2[d2['Customer_id']==3])

         City_Name  Customer_id
329  New York City         3422 

             City_Name  Customer_id
0             Aberdeen            3
17   Arlington Heights            3
22       Atlantic City            3
110            Danbury            3
112              Davis            3
140             Elyria            3
185         Hagerstown            3
227          Kissimmee            3
417         Romeoville            3
441    San Luis Obispo            3
463         Springdale            3


#### People from New York City City  took up the highest number of  subscription 

### Extra Analysis

### Employment rate

In [153]:
job = data.loc[(data['job'] !=  'unknown') & (data['job']!='unemployed')&(data['job']!='student')&(data['job']!='retired') ]
job = job.groupby(['job']).size().to_frame().reset_index().rename(columns = {0 : 'size'})
job

job  size
0         admin.  8981
1    blue-collar  7690
2   entrepreneur  1218
3      housemaid   875
4     management  2463
5  self-employed  1209
6       services  3321
7     technician  5783

In [154]:
unemployed_percentage = job.iloc[-1][1]/job['size'].sum()
employed_percentage = 1 - unemployed_percentage
employment = [round(unemployed_percentage,2), round(employed_percentage,2)]
employment

[0.18, 0.82]

#### We can observe that the percentage of unemployed clients in this bank is very low, less then 18%!

In [155]:
## Subscription Rate by Age

lst = [data]
for column in lst:
    column.loc[column["age"] < 30,  'age_group'] = 20
    column.loc[(column["age"] >= 30) & (column["age"] <= 39), 'age_group'] = 30
    column.loc[(column["age"] >= 40) & (column["age"] <= 49), 'age_group'] = 40
    column.loc[(column["age"] >= 50) & (column["age"] <= 59), 'age_group'] = 50
    column.loc[column["age"] >= 60, 'age_group'] = 60
lst

[       Customer_id  age          job  marital            education  default  \
 0                1   56     services  married          high.school       no   
 1                2   45     services  married             basic.9y  unknown   
 2                3   59       admin.  married  professional.course       no   
 4                5   24   technician   single  professional.course       no   
 5                6   25     services   single          high.school       no   
 ...            ...  ...          ...      ...                  ...      ...   
 37079        37080   73      retired  married  professional.course       no   
 37080        37081   46  blue-collar  married  professional.course       no   
 37081        37082   56      retired  married    university.degree       no   
 37082        37083   44   technician  married  professional.course       no   
 37083        37084   74      retired  married  professional.course       no   
 
       housing loan Region_Code State_

In [156]:
### Subscription Rate by Age 
count_age_response_pct = pd.crosstab(data['y'],data['age_group']).apply(lambda x: x/x.sum() * 100)
count_age_response_pct = count_age_response_pct.transpose() 


age = pd.DataFrame(data['age_group'].value_counts(),index=[30,40,50,20,60])

age['% Contacted'] = age['age_group']*100/age['age_group'].sum()
age['% Subscription'] = count_age_response_pct['yes']
age['% Not Subscription'] = count_age_response_pct['no']
age.drop('age_group',axis = 1,inplace = True)

age['age'] = [30,40,50,20,60]
age = age.sort_values('age',ascending = True)
age

% Contacted  % Subscription  % Not Subscription  age
20    13.751743       15.839493           84.160507   20
30    41.908690       10.207900           89.792100   30
40    25.304949        7.907724           92.092276   40
50    16.356877       10.316051           89.683949   50
60     2.677742       39.696312           60.303688   60

#### Age Stage: Customer who are older are more likely to subscribe the term deposit because they need to prepare for their future life without income after retired.

##### People aged above 60 took up the greatest percentage in subscription, with around 40%. And young people under 20 also contributed a lot, while the middle aged only contributed less than half of the subscription. The main reason behind this situation is that old people always have a strong desire to save money after retirement, while young people have very limited disposable income.
